primitives/src/subset.rs

```diff
@@ -44,30 +45,54 @@ pub trait LatticeOrd {
 /// The type of subsets of an open set of elements of type `A` where the whole set is always
 /// considered to be bigger than any finite set of its elements. This is true for infinite
 /// sets. When talking about finite sets, we have to add that they are _open_.
-///
-/// The meaning of the parameter:
-/// - `None`: no restrictions, the whole set.
-/// - `Some(elems)`: the subset of elements in `elems`.
-///    A particular case is the empty set where `elems` is empty.
-#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Default)]
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
 #[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
-pub struct SubsetRestriction<A: Ord>(pub Option<BTreeSet<A>>);
+pub enum SubsetRestriction<A: Ord> {
+    /// No restrictions, the whole set.
+    Whole,
```

Comment:
I wonder if we should get rid of `Whole` and have it always implied by `Except(None)`

Reply:
Per private discussion, we'll add a separate task to reconsider this since it probably requires storage migration...

Meanwhile, some notes:

1. Using `Except(BTreeSet::new())` is less storage efficient, as you need `[0, 0]`, first encoding the variant of `These` and then the fact that the length is `0`. Meanwhile, `Whole` just requires `[0]` since you encode the tag and then there's no data. There is no change in data size efficiency in Rust terms, as there aren't enough niches to optimize below `32` bytes in either case.
2. Storage migration is likely needed as the data encoding, per above, has changed.

Reply:
Filed as https://polymath.atlassian.net/browse/MESH-1624.